episodes/02-basics.Rmd

```diff
@@ -0,0 +1,330 @@
+numeric types, and provides an initial value in each case.
+```fortran
+program example1
+
+   integer :: i = 1           ! An integer
+   real    :: a = 2.0         ! A floating point number
+   complex :: z = (0.0, 1.0)  ! A complex number as (real-part, imag-part)
+
+end program example1
+```
+Initial values are optional, and these values may be changed later within the code. 
+If a declaration does not specify an initial value, the variable is said to be _undefined_.
+
+### Variable names
+
+The valid Fortran character set for names is `a-z`, `A-Z`, `0-9` and the
+underscore `_`. Valid names must begin with a character. The maximum
+length of a name is 63 characters (introduced in the F2003 standard) with 
+no spaces allowed. This includes names for programs, modules, subroutines, 
+and functions, as well as names for variables.
+
+### The `implicit` statement
+
+By default, variables in fortran do not need to be declared a specific type, they 
+can just be used within the code. Variables with names beginning with letters `i-n` are implicitly
+of type `integer`, while anything else is of type `real` (unless explicitly declared otherwise).
+
+**This is very bad practice and modern Fortran should not be used in this way.** 
+
+The solution to prevent errors involving undeclared variables (usually arising from
+typos) is to declare that no names have implicit type via the use of the
+```fortran
+implicit none
+```
+statment at the very top of the code. 
+
+With this statement inserted, all variable names must be declared explicitly before they
+are referenced. It is still common to see variables beginning with `i-n` as integers.
+
+::::::::::::::::::::::::::::::::::::: challenge
+
+## Use `implicit none`
+
+Edit the above example to include an `implicit none` statment, and print out the three variables
+
+:::::::::::::::: solution
+
+```fortran
+program solution1
+   implicit none
+
+   integer :: i = 1            ! An integer
+   real    :: a = 2.0          ! A floating point number
+   complex :: z = (0.0, 1.0)   ! A complex number as (real-part, imag-part)
+
+   print *, i, a, z
+
+end program solution1
+```
+Compiling and running this code will give the following output
+```
+$ ./a.out 
+  1  2.00000000  (0.00000000,1.00000000)
+```
+
+:::::::::::::::::::::::::
+:::::::::::::::::::::::::::::::::::::::::::::::
+
+### Parameters
+
+In the above example, it is possible to change the values of initialised variables, e.g.
+```fortran
+program example2
+   implicit none
+
+   real :: a = 2.0
+
+   print *, a
+
+   a = a + 5.0
+
+   print *, a
+
+end program example2
+```
+This will then give the output:
+```
+$ ./a.out 
+   2.00000000    
+   7.00000000   
+```
+However, you can also define constant values that do not change by defining variables using `parameter`, e.g.
+```fortran
+program example3
+   implicit none
+
+   real, parameter :: pi = 3.14159265
+
+end program example3
```

Comment:
I personally think it would be useful to either expand this example to show the (compile-time) failure if you try to modify `pi` or leave this is an exercise so users can cause the error themselves.

Reply:
Good point - I'll work on this.